### PR TITLE
Avoid importing all models when instantiating a pipeline

### LINKED ISSUE #24903: Xformers is not installed correctly.
### System Info

- `transformers` version: 4.30.2
- Platform: Linux-5.15.0-76-generic-x86_64-with-glibc2.35
- Python version: 3.10.6
- Huggingface_hub version: 0.16.4
- Safetensors version: 0.3.1
- PyTorch version (GPU?): 2.0.1+cu117 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: yes
- Using distributed or parallel set-up in script?: no

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

``` python 
from transformers import pipeline
pipe = pipeline("text-classification", model="roberta-base", device=0)
```

Edit: I know this model isn't trained for the "text-classification" task, I get the same problem with a private model I fine tuned.

Results in the message

```
...
Xformers is not installed correctly. If you want to use memory_efficient_attention to accelerate training use the following command to install Xformers
pip install xformers.
```

But I'm using torch==2.0.1 and [memory-efficient-attention](https://huggingface.co/docs/diffusers/optimization/fp16#memory-efficient-attention ) states "If you have PyTorch 2.0 installed, you shouldn’t use xFormers!"

The message is confusing - I have torch 2.0 installed and pipeline is for inference. This message doesn't occur if I use `AutoModelForSequenceClassification.from_pretrained`

### Expected behavior

The documentation or the warning message are inconsistent.

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

The check on the model being in the right mapping for the pipeline requires the import of all models associated with that pipeline. This is because it uses the `MODEL_XXX_MAPPING` instead of the `MODEL_XXX_MAPPING_NAMES`. This PR suggests to switch to that, since the check is done on the model name anyway, to avoid weird warnings unrelevant to the user (for instance #24903 gives one example, I also end up having some logs of CUDA kernel being built on my side).

The only feature of the `MODEL_XXX_MAPPING` we need to keep compared to `MODEL_XXX_MAPPING_NAMES` is the extra content (coming from model on the Hub), so this PR adds a reference from the mapping names to the mapping.

Fixes #24903

### CODE REVIEW

File: src/transformers/pipelines/image_segmentation.py [hunk 1]
```diff
@@ -71,14 +71,11 @@ def __init__(self, *args, **kwargs):
...
         requires_backends(self, "vision")
-        self.check_model_type(
-            dict(
-                MODEL_FOR_IMAGE_SEGMENTATION_MAPPING.items()
-                + MODEL_FOR_SEMANTIC_SEGMENTATION_MAPPING.items()
-                + MODEL_FOR_INSTANCE_SEGMENTATION_MAPPING.items()
-                + MODEL_FOR_UNIVERSAL_SEGMENTATION_MAPPING.items()
-            )
-        )
+        mapping = MODEL_FOR_IMAGE_SEGMENTATION_MAPPING_NAMES.copy()
```
Comment (sgugger) [inline src/transformers/pipelines/image_segmentation.py:74]: The sum is not working anymore on the ordered dicts, so had to write it this way.

File: src/transformers/pipelines/object_detection.py [hunk 2]
```diff
@@ -53,9 +56,9 @@ def __init__(self, *args, **kwargs):
             raise ValueError(f"The {self.__class__} is only available in PyTorch.")
 
         requires_backends(self, "vision")
-        self.check_model_type(
-            dict(MODEL_FOR_OBJECT_DETECTION_MAPPING.items() + MODEL_FOR_TOKEN_CLASSIFICATION_MAPPING.items())
-        )
+        mapping = MODEL_FOR_OBJECT_DETECTION_MAPPING_NAMES.copy()
+        mapping.update(MODEL_FOR_TOKEN_CLASSIFICATION_MAPPING_NAMES)
```
Comment (sgugger) [inline src/transformers/pipelines/object_detection.py:60]: I didn't change the type but token classification here seems like a mistake?

**Review by Narsil — APPROVED** [review]
LGTM.

This PR is actually removing all the checks directly: 
https://github.com/huggingface/transformers/pull/24827

Removing the checks allows a more streamline `pipeline(model=MyModel())`.

Maybe just removing the checks here works too ?

**Comment by sgugger** [comment]
It works too, but I think with the design selected for the PEFT integration in `from_pretrained` (which will supercede #24827), we won't need to remove them. I think it's a useful warning to have when users try to instantiate a pipeline with a model not suitable for it.

**Review by LysandreJik — APPROVED** [review]
That's much better! Thanks for investigating, should make pipeline loading much better.

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 3]
```diff
@@ -205,7 +205,7 @@ def __init__(
 
         if self.model.config.model_type == "whisper":
             self.type = "seq2seq_whisper"
-        elif self.model.__class__ in MODEL_FOR_SPEECH_SEQ_2_SEQ_MAPPING.values():
+        elif self.model.__class__.__name__ in MODEL_FOR_SPEECH_SEQ_2_SEQ_MAPPING_NAMES.values():
```
Comment (fxmarty) [inline src/transformers/pipelines/automatic_speech_recognition.py:208]: @sgugger @ArthurZucker Am I correct that this breaks if an user registered a model with `AutoModelForSpeechSeq2Seq.register(AutoConfig, MyCustomModel)`?

Comment (ArthurZucker) [reply src/transformers/pipelines/automatic_speech_recognition.py:208]: From the description of the PR I would think that no (`_extra_content` is what should have been kept).
The `self._model_mapping._model_mapping = self` adds a field to `MODEL_FOR_SPEECH_SEQ_2_SEQ_MAPPING_NAMES` but it's not taken into account when computing the `values()` if I am not mistaken

Comment (sgugger) [reply src/transformers/pipelines/automatic_speech_recognition.py:208]: No new key is added `MODEL_FOR_SPEECH_SEQ_2_SEQ_MAPPING_NAMES`, but you can check for its `_model_mapping` if it exists, as is done in the diff below in `pipelines/base.py`

**Pull request merged**